puts/harness/CMakeLists.txt

```diff
@@ -0,0 +1,62 @@
+cmake_minimum_required(VERSION 3.15)
+
+project(puts VERSION 0.1 LANGUAGES C)
+
+enable_testing()
+
+set(PUTS "")
+set(PUT_LIBS "")
+
+string(REPLACE "," ";" VENDORS "${VENDORS}")
+foreach(vendor_path IN LISTS VENDORS)
+    set(PROTOCOL "tls")
+    set(HARNESS "openssl")
+
+    # FIXME retrieve mk_vendor config uid if available
+    string(SHA256 VENDOR_HASH "${PROTOCOL}/${HARNESS}/${vendor_path}")
+    string(SUBSTRING ${VENDOR_HASH} 0 16 VENDOR_HASH)
+
+    set(PUT_ID "${VENDOR_HASH}")
+
+    add_library(put-${PROTOCOL}-${HARNESS}-${PUT_ID} OBJECT ${PROTOCOL}/${HARNESS}/src/put.c)
+    set_property(TARGET put-${PROTOCOL}-${HARNESS}-${PUT_ID} PROPERTY C_VISIBILITY_PRESET default)
+    target_compile_definitions(put-${PROTOCOL}-${HARNESS}-${PUT_ID} PRIVATE PUT_ID=${PUT_ID} REGISTER=register_${HARNESS}_${PUT_ID})
+    target_include_directories(put-${PROTOCOL}-${HARNESS}-${PUT_ID} PRIVATE "include")
+    target_include_directories(put-${PROTOCOL}-${HARNESS}-${PUT_ID} PRIVATE "${vendor_path}/include")
+
+    add_custom_command(
+        COMMAND clang -flto -nostdlib -Wl,-r $<TARGET_OBJECTS:put-${PROTOCOL}-${HARNESS}-${PUT_ID}> "-L${vendor_path}/lib" -lssl -lcrypto -o ${CMAKE_BINARY_DIR}/put-${PROTOCOL}-${HARNESS}-${PUT_ID}.o OUTPUT ${CMAKE_BINARY_DIR}/put-${PROTOCOL}-${HARNESS}-${PUT_ID}.o DEPENDS put-${PROTOCOL}-${HARNESS}-${PUT_ID}
+        COMMENT "relocation for put-${PROTOCOL}-${HARNESS}-${PUT_ID}" VERBATIM)
+
+    add_custom_target(relocate-put-${PROTOCOL}-${HARNESS}-${PUT_ID} ALL DEPENDS ${CMAKE_BINARY_DIR}/put-${PROTOCOL}-${HARNESS}-${PUT_ID}.o)
+
+    add_library(relocated-put-${PROTOCOL}-${HARNESS}-${PUT_ID} OBJECT IMPORTED)
+    set_property(TARGET relocated-put-${PROTOCOL}-${HARNESS}-${PUT_ID} PROPERTY IMPORTED_OBJECTS ${CMAKE_BINARY_DIR}/put-${PROTOCOL}-${HARNESS}-${PUT_ID}.o)
+    add_dependencies(relocated-put-${PROTOCOL}-${HARNESS}-${PUT_ID} relocate-put-${PROTOCOL}-${HARNESS}-${PUT_ID})
+
+    list(APPEND PUTS "${HARNESS}_${PUT_ID}")
+    list(APPEND PUT_LIBS relocated-put-${PROTOCOL}-${HARNESS}-${PUT_ID})
+endforeach()
+
+set(BUNDLE_OBJECTS "")
+foreach(PUT_LIB IN LISTS PUT_LIBS)
+    list(APPEND BUNDLE_OBJECTS "$<TARGET_OBJECTS:${PUT_LIB}>")
+    install(FILES "$<TARGET_OBJECTS:${PUT_LIB}>" DESTINATION ".")
+endforeach()
+
+add_library(puts-bundle STATIC ${BUNDLE_OBJECTS})
+set_target_properties(puts-bundle PROPERTIES LINKER_LANGUAGE C)
+
+# create Rust init file
+set(RUST_INIT_FILE ${CMAKE_BINARY_DIR}/init.rs)
+set(RUST_INIT_DECL "")
+set(RUST_INIT_CALL "")
+foreach(put_id IN LISTS PUTS)
+    string(APPEND RUST_INIT_DECL "    fn register_${put_id}(callback: FnRegister);\n")
+    string(APPEND RUST_INIT_CALL "        register_${put_id}(callback);\n")
+endforeach()
+
+file(WRITE ${RUST_INIT_FILE} "use crate::FnRegister;\n\nextern {\n${RUST_INIT_DECL}}\n\npub fn register(callback: FnRegister) {\n    unsafe {\n${RUST_INIT_CALL}    }\n}\n")
```

Comment:
Note to myself: Looks interesting, might need to go back to this.

Reply:
This is what will allow us to register the C PUTs in tlspuffin in the future. We generate a tiny Rust file performing just the PUT registration for each of the PUT, and this is included into the `tlspuffin-harness-sys`crate. It can then be called with specific callbacks when creating a `PutRegistry`.

For the record, we first planned to do that using compiler facilities like `__attribute__((constructor))` and made it work. However, I found several mentions that this have been broken in Rust for some versions of clang and that it might not be the way forward that Rust choose to support. Using code generation is future-proof and hopefully more consistent with the way the Rust ecosystem works (e.g. bindgen).

I'm not particularly proud of this specific cmake code though. I intend to change this into a template of sort once I figure out the best way to do that.